Convert block ptr type layouts to `Subgroup2DBlockEncoding` layouts

alexbaden opened the pull request:
To ensure Subgroup2DBlockIO layouts survive RemoveLayoutConversions we need to identify them as anchor layouts on the `LoadOp`. To do that, we need to change the load layout to be a `Subgroup2DBlockIO` layout and propagate that layout as necessary through the existing IR. This PR introduces `TritonIntelGPU::OptimizeBlockEncoding` which replaces the encoding for all ptr types between the candidate `LoadOp` and the `MakeTensorPtr` Op which creates the ptr for the load. The pass is based on a similar pass from upstream, `TritonNVIDIAGPU::OptimizeDescriptorEncoding`, which modifies layouts in-place for tensor descriptors. To avoid modifying any non-ptr types, which should be modified using `ConvertLayoutOp`, we add a dummy layout conversion after the `LoadOp` which will be removed when the old load layout is removed in `RemoveLayoutConversions`. 

With this pass we are able to create the desired chain of `Subgroup2DBlockIO` -> `ConvertLayout` -> `DPAS` which will allow us to begin using `Subgroup2DBlockIO` in the LLVM lowering and drive the layout conversion via LinearLayout objects. The pass is introduced in this PR unused, but I did include a lit test which covers the most common for loop use case. I intend to improve test coverage as we expand use of the pass. 

~Depends on #4461~

Close #4362

alexbaden committed:
Add pass to convert block load to subgroup 2d block encoding types

alexbaden commented on third_party/intel/backend/compiler.py:
```diff
@@ -290,6 +290,7 @@ def make_ttgir(mod, metadata, opt, properties):
 
         intel.passes.ttgpuir.add_accelerate_matmul(pm)
         intel.passes.ttgpuir.add_materialize_block_pointer(pm)
+        intel.passes.ttgpuir.add_optimize_block_load_encoding(pm)
```
Will remove this line after I get the first CI run (or make it off behind a flag).

alexbaden replied:
Actually, since the subgroup2d block layout is not an anchor layout this pass becomes a no-op. I will run the benchmarks but I think this can be left on.

alexbaden commented:
Benchmarks unchanged - as expected, this should be a no-op.
https://benchmarks.glados.intel.com/d/1pXX4hUSz/microbenchmarks?orgId=1&var-tag=ci%7Cpr4463a&var-table=ci&var-bench=All&var-device=All&var-compiler=All&var-backend=All&var-baseline_backend=cutlass-ci-Max%201550&var-target_backend=cutlass-ci-Max%201550&var-okr_device=All

etiotto commented on third_party/intel/include/Dialect/TritonIntelGPU/Transforms/Passes.td:
```diff
@@ -409,4 +409,15 @@ def TritonIntelGPUReduceVariableLiveness
                            "mlir::arith::ArithDialect"];
 }
+
+def TritonIntelGPUOptimizeBlockIOEncodingPass : Pass<"tritonintelgpu-optimize-block-io-encoding", "mlir::ModuleOp"> {
```
[nit]: drop "Pass" suffix from the transformation name.

etiotto commented on third_party/intel/include/Dialect/TritonIntelGPU/Transforms/Passes.td:
```diff
@@ -409,4 +409,15 @@ def TritonIntelGPUReduceVariableLiveness
+def TritonIntelGPUOptimizeBlockIOEncodingPass : Pass<"tritonintelgpu-optimize-block-io-encoding", "mlir::ModuleOp"> {
+  let summary = "Set encodings on candidates for Subgroup 2D Block IO ops";
+
+  let description = [{
```
Question:  `tt.load` operations feeding a `tt.dot` operation are expected to have "dot" encoding in their type. How does this work with the new encoding ? Will we need a layout conversion operation to change the layout from "BlockEncoding" to "Dot" encoding ?

etiotto commented on third_party/intel/lib/TritonIntelGPUTransforms/OptimizeBlockIOEncoding.cpp:
```diff
@@ -0,0 +1,319 @@
+namespace ttg = mlir::triton::gpu;
+namespace ttgi = mlir::triton::gpu::intel;
+
+namespace mlir {
+namespace triton {
+namespace gpu::intel {
```
```suggestion
namespace mlir::triton::gpu::intel {
```

alexbaden replied:
why?

etiotto commented on third_party/intel/lib/TritonIntelGPUTransforms/OptimizeBlockIOEncoding.cpp:
```diff
@@ -0,0 +1,319 @@
+class TritonIntelGPUOptimizeBlockIOEncodingPass
+    : public impl::TritonIntelGPUOptimizeBlockIOEncodingPassBase<
+          TritonIntelGPUOptimizeBlockIOEncodingPass> {
+
```
[Suggestion]: Can you put the public members at the top and the private members at the bottom of the class. That way the public interface is upfront.

alexbaden replied:
It seemed like upstream passes put `runOnOperation` at the bottom of the file by convention - so when you want to know what a pass does, you start by scrolling all the way down and read bottom to top. I tried to follow that, but if we want to standardize and write down a different convention I think that would be fine.

etiotto commented on test/TritonIntelGPU/optimize-block-io-encoding.mlir:
```diff
@@ -0,0 +1,65 @@
+// CHECK: #mma1 = #ttig.subgroup_2d_block<{warpsPerCTA = [8, 4], instrShape = [16, 16], numBlocks=2, order=[0, 1], kWidth=2, threadsPerWarp=16}>
+// CHECK: #mma2 = #ttig.dpas<{repeatCount = 8, systolicDepth = 8, executionSize = 16, opsPerChan = 2, threadsPerWarp = 16, warpsPerCTA = [8, 4], repCluster = [4, 2], A = [32, 16], B = [16, 32], C = [32, 32]}>
+#mma = #ttig.dpas<{repeatCount = 8, systolicDepth = 8, executionSize = 16, opsPerChan = 2, threadsPerWarp = 16, warpsPerCTA = [8, 4], repCluster = [4, 2], A = [32, 16], B = [16, 32], C = [32, 32]}>
+module attributes {"ttg.num-ctas" = 1 : i32, "ttg.num-warps" = 32 : i32, ttg.target = "xpu", "ttg.threads-per-warp" = 16 : i32, ttig.min_sg_size = 16 : i32, ttig.support_bf16_conversion, ttig.support_dpas, ttig.support_sg_2d_block, ttig.target_arch = "spir64"} {
```
Can the module declaration be simplified to: 
`module attributes {"ttg.num-ctas" = 1 : i32, "ttg.num-warps" = 32 : i32`} {`

alexbaden replied:
No. I removed some of the ones but left a combination of attributes that I believe are relevant and attributes that are required.

etiotto commented on test/TritonIntelGPU/optimize-block-io-encoding.mlir:
```diff
@@ -0,0 +1,65 @@
+// CHECK: #mma2 = #ttig.dpas<{repeatCount = 8, systolicDepth = 8, executionSize = 16, opsPerChan = 2, threadsPerWarp = 16, warpsPerCTA = [8, 4], repCluster = [4, 2], A = [32, 16], B = [16, 32], C = [32, 32]}>
+#mma = #ttig.dpas<{repeatCount = 8, systolicDepth = 8, executionSize = 16, opsPerChan = 2, threadsPerWarp = 16, warpsPerCTA = [8, 4], repCluster = [4, 2], A = [32, 16], B = [16, 32], C = [32, 32]}>
+module attributes {"ttg.num-ctas" = 1 : i32, "ttg.num-warps" = 32 : i32, ttg.target = "xpu", "ttg.threads-per-warp" = 16 : i32, ttig.min_sg_size = 16 : i32, ttig.support_bf16_conversion, ttig.support_dpas, ttig.support_sg_2d_block, ttig.target_arch = "spir64"} {
+  tt.func public @matmul_kernel_with_block_pointers(%arg0: !tt.ptr<f16>, %arg1: !tt.ptr<f16>, %arg2: !tt.ptr<f16>) attributes {noinline = false} {
```
Remove `attributes {noinline = false}` as it is not strictly necessary

etiotto commented on test/TritonIntelGPU/optimize-block-io-encoding.mlir:
```diff
@@ -0,0 +1,65 @@
+    %11 = arith.muli %8, %c256_i32 : i32
+    // CHECK: tt.make_tensor_ptr {{.*}} : <tensor<32x256xf16, #mma1>>
+    %12 = tt.make_tensor_ptr %arg1, [%c5120_i64, %c4096_i64], [%c4096_i64, %c1_i64], [%c0_i32, %11] {order = array<i32: 1, 0>} : <tensor<32x256xf16, #blocked2>>
+    %13:3 = scf.for %arg3 = %c0_i32 to %c5120_i32 step %c32_i32 iter_args(%arg4 = %cst, %arg5 = %10, %arg6 = %12) -> (tensor<256x256xf32, #blocked>, !tt.ptr<tensor<256x32xf16, #blocked1>>, !tt.ptr<tensor<32x256xf16, #blocked2>>)  : i32 {
```
Add a CHECK to ensure that the loop init args have the new tensor ptr type with %mma layout.
The return types of the `scf.for` operation should be CHECKED too.

alexbaden replied:
The return type of the scf is not modified - so I am not sure what we need to check? The `tt.dot` return type is unchanged and the make tensor ptr / store uses the same layout as before. 
```
 %26 = tt.dot %24, %25, %23, inputPrecision = tf32 : tensor<256x32xf16, #ttg.dot_op<{opIdx = 0, parent = #mma2, kWidth = 1}>> * tensor<32x256xf16, #ttg.dot_op<{opIdx = 1, parent = #mma2, kWidth = 2}>> -> tensor<256x256xf32, #mma2>
      %27 = ttg.convert_layout %26 : tensor<256x256xf32, #mma2> -> tensor<256x256xf32, #blocked>
      %28 = tt.advance %arg5, [%c0_i32, %c32_i32] : <tensor<256x32xf16, #mma>>
      %29 = tt.advance %arg6, [%c32_i32, %c0_i32] : <tensor<32x256xf16, #mma1>>
      scf.yield %27, %28, %29 : tensor<256x256xf32, #blocked>, !tt.ptr<tensor<256x32xf16, #mma>>, !tt.ptr<tensor<32x256xf16, #mma1>>
    }
    %14 = tt.make_tensor_ptr %arg2, [%c1024_i64, %c4096_i64], [%c4096_i64, %c1_i64], [%9, %11] {order = array<i32: 1, 0>} : <tensor<256x256xf16, #blocked2>>
    %15 = arith.truncf %13#0 : tensor<256x256xf32, #blocked> to tensor<256x256xf16, #blocked>
    %16 = ttg.convert_layout %15 : tensor<256x256xf16, #blocked> -> tensor<256x256xf16, #blocked2>
    tt.store %14, %16 {boundaryCheck = array<i32: 0, 1>} : !tt.ptr<tensor<256x256xf16, #blocked2>>
    tt.return
```
I can add a check to make sure the arguments of the scf for loop are using the new values, but the verifier for the scf for should not allowed a type mismatch so I don't think we need another type check (I briefly looked through other lit examples and did not see arg type checking for `scf.for`).

etiotto commented on test/TritonIntelGPU/optimize-block-io-encoding.mlir:
```diff
@@ -0,0 +1,65 @@
+      %26 = tt.advance %arg5, [%c0_i32, %c32_i32] : <tensor<256x32xf16, #blocked1>>
+      // CHECK: tt.advance {{.*}} : <tensor<32x256xf16, #mma1>>
+      %27 = tt.advance %arg6, [%c32_i32, %c0_i32] : <tensor<32x256xf16, #blocked2>>
+      scf.yield %25, %26, %27 : tensor<256x256xf32, #blocked>, !tt.ptr<tensor<256x32xf16, #blocked1>>, !tt.ptr<tensor<32x256xf16, #blocked2>>
```
The scf.yield operand block pointer operands should have #mma layout, can you add a CHECK here ?

etiotto commented on test/TritonIntelGPU/optimize-block-io-encoding.mlir:
```diff
@@ -0,0 +1,65 @@
+      scf.yield %25, %26, %27 : tensor<256x256xf32, #blocked>, !tt.ptr<tensor<256x32xf16, #blocked1>>, !tt.ptr<tensor<32x256xf16, #blocked2>>
+    }
+    %14 = tt.make_tensor_ptr %arg2, [%c1024_i64, %c4096_i64], [%c4096_i64, %c1_i64], [%9, %11] {order = array<i32: 1, 0>} : <tensor<256x256xf16, #blocked2>>
+    %15 = arith.truncf %13#0 : tensor<256x256xf32, #blocked> to tensor<256x256xf16, #blocked>
```
Add CHECK to verify that %13#0 has the dpas layout.

etiotto commented on third_party/intel/lib/TritonIntelGPUTransforms/OptimizeBlockIOEncoding.cpp:
```diff
@@ -0,0 +1,319 @@
+namespace {
+
+SmallVector<Value> getTiedArgs(Operation *op, int resultIdx) {
+  if (auto forOp = dyn_cast<scf::ForOp>(op)) {
```
The code for `scf.for` and `scf.while` can be commoned by dynamic casting `op` to a `LoopLikeOpInterface`, which both operations implement.

whitneywhtsang replied:
`LoopLikeOpInterface` should be get loop info with common name.

etiotto commented on third_party/intel/lib/TritonIntelGPUTransforms/OptimizeBlockIOEncoding.cpp:
```diff
@@ -0,0 +1,319 @@
+    auto yieldVal =
+        whileOp.getBeforeBody()->getTerminator()->getOperand(resultIdx);
+    auto initVal = whileOp.getOperands()[resultIdx];
+    return {iterArg, result, iterArg, initVal};
```
After a return there is no need for and `else if`, can you pls change it to an if.

etiotto reviewed (commented):
Can we add lit test that exercise and scf.while loop and a if statement  as well.

alexbaden commented:
Hi @etiotto we just discussed in the architecture meeting refactoring `SmallVector<Value> getTiedArgs(Operation *op, int resultIdx)` into a common utility since it is duplicated from the upstream pass - https://github.com/triton-lang/triton/blob/main/lib/Dialect/TritonNvidiaGPU/Transforms/OptimizeDescriptorEncoding.cpp#L131. So, I would prefer not to make cosmetic changes that might be controversial getting merged upstream.

alexbaden commented on test/TritonIntelGPU/optimize-block-io-encoding.mlir:
```diff
@@ -0,0 +1,65 @@
+    tt.store %14, %16 {boundaryCheck = array<i32: 0, 1>} : !tt.ptr<tensor<256x256xf16, #blocked2>>
+    tt.return
+  }
+}
```
I think it is probably failing because the advance occurs before the load and the argument to advance is not `MakeTensorPtrOp` but the while loop construction. Why do you expect this to work? Can this ever be generated from a Triton kernel  written in Python? I have run this PR on all the tests and benchmarks and not seen this error - so if it legal then we are missing unit test coverage for this pattern.

alexbaden committed:
validate scf for loop type changes

alexbaden committed:
reduce set of module attributes

alexbaden commented on test/TritonIntelGPU/optimize-block-io-encoding.mlir:
```diff
@@ -0,0 +1,65 @@
+// RUN: triton-opt %s -split-input-file --allocate-shared-memory --tritonintelgpu-optimize-block-io-encoding | FileCheck %s
```
removed

whitneywhtsang commented on test/TritonIntelGPU/optimize-block-io-encoding.mlir:
```diff
@@ -0,0 +1,67 @@
+#blocked = #ttg.blocked<{sizePerThread = [4, 4], threadsPerWarp = [1, 16], warpsPerCTA = [8, 4], order = [1, 0]}>
+#blocked1 = #ttg.blocked<{sizePerThread = [1, 8], threadsPerWarp = [4, 4], warpsPerCTA = [32, 1], order = [1, 0]}>
+#blocked2 = #ttg.blocked<{sizePerThread = [1, 8], threadsPerWarp = [1, 16], warpsPerCTA = [16, 2], order = [1, 0]}>
+// CHECK: #mma = #ttig.subgroup_2d_block<{warpsPerCTA = [8, 4], instrShape = [8, 16], numBlocks=2, order=[1, 0], kWidth=1, threadsPerWarp=16}>
+// CHECK: #mma1 = #ttig.subgroup_2d_block<{warpsPerCTA = [8, 4], instrShape = [16, 16], numBlocks=2, order=[0, 1], kWidth=2, threadsPerWarp=16}>
+// CHECK: #mma2 = #ttig.dpas<{repeatCount = 8, systolicDepth = 8, executionSize = 16, opsPerChan = 2, threadsPerWarp = 16, warpsPerCTA = [8, 4], repCluster = [4, 2], A = [32, 16], B = [16, 32], C = [32, 32]}>
```
?
```suggestion
// CHECK-DAG: [[2DBLOCK_A:#.*]] = #ttig.subgroup_2d_block<{warpsPerCTA = [8, 4], instrShape = [8, 16], numBlocks=2, order=[1, 0], kWidth=1, threadsPerWarp=16}>
// CHECK-DAG: [[2DBLOCK_B:#.*]] = #ttig.subgroup_2d_block<{warpsPerCTA = [8, 4], instrShape = [16, 16], numBlocks=2, order=[0, 1], kWidth=2, threadsPerWarp=16}>
// CHECK-DAG: [[DPAS:#.*]] = #ttig.dpas<{repeatCount = 8, systolicDepth = 8, executionSize = 16, opsPerChan = 2, threadsPerWarp = 16, warpsPerCTA = [8, 4], repCluster = [4, 2], A = [32, 16], B = [16, 32], C = [32, 32]}>
```

alexbaden replied:
The ordering should be consistent as the dot op operands are processed in order. I think we would want to know if the order somehow changed. But, having variable names for the layouts is much easier to read.

alexbaden replied:
Decided to update to CHECK-DAG when I added the complex test.

whitneywhtsang commented on third_party/intel/lib/TritonIntelGPUTransforms/OptimizeBlockIOEncoding.cpp:
```diff
@@ -0,0 +1,319 @@
+
+    PointerType oldType = val.getType();
+    auto oldTensorTy = cast<RankedTensorType>(oldType.getPointeeType());
+    auto newTensorTy = RankedTensorType::get(
```
why not call getNewType?

alexbaden replied:
Looks like I forgot those were there :) 

This only happens once in the function so I am inclined to delete the free function definitions and just convert here - it's much clearer what is changing, for one thing, then having to check the function body above. What do you think?

whitneywhtsang commented on third_party/intel/lib/TritonIntelGPUTransforms/OptimizeBlockIOEncoding.cpp:
```diff
@@ -0,0 +1,319 @@
+    auto newTensorTy = RankedTensorType::get(
+        oldTensorTy.getShape(), oldTensorTy.getElementType(), newEncoding);
+
+    val.setType(PointerType::get(newTensorTy, oldType.getAddressSpace()));
```
why not call getNewPointerType?

alexbaden replied:
deleted

whitneywhtsang commented on third_party/intel/lib/TritonIntelGPUTransforms/OptimizeBlockIOEncoding.cpp:
```diff
@@ -0,0 +1,319 @@
+                          << itr->second.desiredEncoding << " for value "
+                          << typedVal << ". Ensure new encoding "
+                          << info.desiredEncoding << " matches.\n");
+        assert(itr->second == info && "already visited encoding info for "
```
I think the assertion is not triggered as only a subset of uses kind are considered.

alexbaden replied:
Right, but if a use is missed wouldn't we see the IR fail to verify?

whitneywhtsang commented on third_party/intel/lib/TritonIntelGPUTransforms/OptimizeBlockIOEncoding.cpp:
```diff
@@ -0,0 +1,319 @@
+        }
+        break;
+      }
+      return isa<LoadOp>(v.getDefiningOp()) ? cast<LoadOp>(v.getDefiningOp())
+                                            : nullptr;
```
```suggestion
      return dyn_cast<LoadOp>(v.getDefiningOp());
```

alexbaden committed:
fixup typo, cleanup load cast

alexbaden committed:
break long lines

alexbaden commented:
FYI `getTiedArgs` was refactored upstream into shared utils, so I will remove it and use the shared impl once we merge in that change: https://github.com/triton-lang/triton/commit/4d791f09355ba4e82123e72af29958d588092b3a

alexbaden committed:
remove dead functions

alexbaden committed:
use attr variable in lit test

whitneywhtsang commented on third_party/intel/lib/TritonIntelGPUTransforms/OptimizeBlockIOEncoding.cpp:
```diff
@@ -0,0 +1,319 @@
+      } else if (isa<scf::YieldOp>(op)) {
+        auto vals = getTiedArgs(op->getParentOp(), use.getOperandNumber());
+        updateEncoding(vals, EncodingInfo{encoding});
+      }
```
like what if there is another load uses the same pointer and it doesn't have the row_major attr?

If you don't expect any other kind of operation, should we add a llvm_unreachable?

whitneywhtsang commented on third_party/intel/lib/TritonIntelGPUTransforms/OptimizeBlockIOEncoding.cpp:
```diff
@@ -0,0 +1,319 @@
+      if (isa<scf::ForOp, scf::WhileOp, scf::IfOp>(definingOp)) {
+        auto vals = getTiedArgs(definingOp, opResult.getResultNumber());
+        updateEncoding(vals, EncodingInfo{encoding});
+      }
```
what about there is an advance op between make tensor ptr op and the load opt?

If you don't expect any other kind of operation, should we add a llvm_unreachable?

alexbaden replied:
It shouldn't be unreachable because there might be users that change the type - a load is a good example. I've added `tt.advance` but I can't think of any additional user that would take in a pointer and return the same pointer.

whitneywhtsang replied:
offline discussion: `tt.advance` can be a defining op, but it should be used by `yield` in the usual case, and it can be handled as the user of `yield`.

alexbaden committed:
add complex control flow test

alexbaden committed:
remove unnecessary assert

alexbaden committed:
further reduce unecessary test values

alexbaden committed:
properly handle while loops

alexbaden committed:
propagate layout to tt.advance

alexbaden commented:
Thanks @etiotto and @whitneywhtsang for your comments - I have increased the testing to show more structured control flow examples to include if, while, and for loops. While doing this I discovered a bug in the getTiedArgs function which I fixed, and which is also merged upstream. I also fixed an issue in getMakeTensorPtr for while loops which I have PRed separately (#4567) but included and tested here (I think it would be nice to have that commit be logically separate). Finally, I added support for `tt.advance` in nested control flow. I believe this should address all the remaining comments above.

alexbaden commented:
We need the bug fix too: https://github.com/triton-lang/triton/commit/e71689d7c7db7d5bbbcfe48363c452d184f86e5f

alexbaden commented on third_party/intel/lib/TritonIntelGPUTransforms/OptimizeBlockIOEncoding.cpp:
```diff
@@ -0,0 +1,311 @@
+        oldTensorTy.getShape(), oldTensorTy.getElementType(), newEncoding);
+
+    val.setType(PointerType::get(newTensorTy, oldType.getAddressSpace()));
+    if (einfo.requiresConvert) {
```
The goal of the algorithm in the pass is to attach new encoding info to each value with the initial propagation and then update all values in the second pass. It seemed like `requiresConvert` fit more naturally in the first phase, where the encoding info is being determined and propagated. But, I think you are right that we can look at all users of each Value in the second phase and add the convert if any one of the users is a load. 

I think I prefer storing `requiresConvert` in the first phase - both because it seems to fit more naturally there and because it serves as an additional sanity check on the algorithm (if we see the same Value again, we check to make sure `requiresConvert` matches the last time we saw that Value). I suppose we could store the index of the User in `requiresConvert` instead of making it a boolean, and avoid the second Users loop in phase 2, but the users loop doesn't seem too expensive. I also think we could extend the use of this pass to propagate transpose layouts, too, so having `requiresConvert`, or at least having the machinery to process additional properties, may be useful.

alexbaden commented on third_party/intel/lib/TritonIntelGPUTransforms/OptimizeBlockIOEncoding.cpp:
```diff
@@ -0,0 +1,311 @@
+    // Note: we need the old layout to get the order for the load, but it is not
+    // clear the layout will always be Blocked. Is there a better way to get
+    // this info?
+    auto oldLayout = cast<BlockedEncodingAttr>(oldTensorType.getEncoding());
```
Turns out there was a better way to get that info! :) Removed.

alexbaden committed:
remove unused var

alexbaden commented:
For some reason I can't reply to your `tt.advance` comment but yes, I added explicit processing for `tt.advance` because the result of `tt.advance` was not being updated. The operand was being updated as part of the existing control flow handling. If `tt.advance` was being used by an op other than the `LoadOp` being targeted as part of this pass then we would see a failure, but that failure would be caught during the pass pipeline and up-leveled as a type mismatch which would be pretty easy to detect. In all the benchmarks and unit tests I have run I have not seen such a scenario.

whitneywhtsang commented:
> We need the bug fix too: [triton-lang/triton@e71689d](https://github.com/triton-lang/triton/commit/e71689d7c7db7d5bbbcfe48363c452d184f86e5f) but I don't think we need to hold this PR for those merges.

FYI merged too.

whitneywhtsang reviewed (commented):
ARs captured from the architecture meeting:
- Rebase to utilize the upstream getTiedArgs implementation.
- Remove the `requiresConvert` field.
- Add a lit test for the case where the result of an advance op is utilized by the load op.
- Ensure that the performance of tutorial 6 remains unaffected by these changes.

whitneywhtsang committed:
address review comments
